pep-9999.rst

```diff
@@ -0,0 +1,341 @@
+responsible to ensure resolution of lazy objects on lookup. 
+
+To avoid a performance penalty on the vast majority of dictionaries which never
+contain any lazy object, we install a specialized lookup function
+(``lookdict_unicode_lazy``) for module namespace dictionaries when they first
+gain a lazy-object value. This lookup function checks the looked-up value and
+resolves it before returning it, if it is a lazy object.
+
+This implementation comprehensively prevents leakage of lazy objects, ensuring
+they are always resolved to the real imported object before anyone can get hold
+of them for any use, while avoiding any noticeable performance impact on
+dictionaries in general.
+
+
+Specification
+=============
+
+Lazy imports are opt-in, and globally enabled via a new ``-L`` flag to the
+python interpreter, or a ``PYTHONLAZYIMPORTS`` environment variable.
+
+When enabled, the loading and execution of all (and only) top level imports is
+deferred until the imported name is used. This could happen immediately (e.g.
+on the very next line after the import statement) or much later (e.g. while
+using the name inside a function being called by some other code at some later
+time.)
+
+For these top level imports, there are two exceptions which will make them
+eager (not lazy): imports inside ``try``/``except``/``finally``  or ``with``
+blocks, and star imports (``from foo import *``.) Imports inside
+exception-handling blocks remain eager so that any exceptions arising from the
+import can be handled. Star imports must remain eager since performing the
+import is the only way to know which names should be added to the namespace.
+
+Imports inside class definitions or inside functions/methods are not "top
+level" and are never lazy.
+
+Dynamic imports using ``__import__`` or ``importlib.import_module`` are also
+never lazy.
+
+
+Per-module opt out
+------------------
+
+Due to the backwards compatibility issues mentioned below, it may be necessary
+to force some imports to be eager. In first-party code, this can be easily
+accomplished via any module-level reference to the name, e.g. even re-assigning
+the name to itself will trigger the import:
+
+.. code-block python
+    import foo
+    
+    # ensure 'foo' is eagerly imported
+    foo = foo
+
+Another option that scales better to making multiple imports lazy is to place
+them inside a ``try/finally``:
+
+.. code-block py
+    try:  # force these imports to be eager
+        import foo
+        import bar
+    finally:
+        pass
+
+The more difficult case can occur if an import in third-party code that can't
+easily be modified must be forced to be eager. For this purpose, we propose to
+add an API to ``importlib`` that can be called early in the process to specify
+a list of module names within which all imports will be eager:
+
+.. code-block py
+    from importlib import set_eager_imports
+
+    set_eager_imports(["one.mod", "another"])
```

Comment:
What about a context manager in importlib so that you could do something like:

```python
with importlib.eager():
    # all things inside this with block will be opted out of lazy imports
```

Reply:
I like this idea. @Kronuz was concerned that there might be confusion about whether the effect of `with importlib.eager():` is shallow or deep. I.e. does it make all imports eager that occur recursively "within" that block, or just the imports directly in this module? I prefer shallow, because it's less action-at-a-distance; I'd rather know when reading a module whether the imports in it will be shallow or lazy (presuming lazy imports is enabled at all.) I think if we don't suggest the possibility of "deep" effect then the shallow version should be fairly intuitive...

Reply:
If the effect is shallow, the context manager can literally be a no-op, since as discussed elsewhere in the PEP, context managers automatically make imports within them eager, for exception-handling reasons.

Reply:
I was thinking it would enable deep eager imports because it's hard otherwise to enforce that, but it *could* be controlled by a flag, e.g.:

```python
with importlib.eager(deep=False)
    ...
with importlib.eager(deep=True)
    ...
```

Or `shallow` as the flag name instead of `deep`.   Either way, maybe shallow eagerness should be the default?

Reply:
> If the effect is shallow, the context manager can literally be a no-op, since as discussed elsewhere in the PEP, context managers automatically make imports within them eager, for exception-handling reasons.

Oh yeah, right!

Reply:
I think that the `importlib.set_eager_imports()` API currently proposed in the PEP serves the same use case as `with importlib.eager(deep=True)` does. But I think `set_eager_imports()` is a more predictable and less error-prone (though arguably less aesthetic) way to do it, since it ensures that certain selected modules will always use only eager imports, regardless of where they are imported from, whereas if you use the deep context manager, you never know whether some other module might first import the same thing _without_ the deep-eager context, so you might be confused why that thing still gets imported with laziness initially..

So my inclination would be to provide `importlib.set_eager_imports()` but not provide a deep context manager.

Reply:
I can see that.  The one thing that maybe bothers me is that `set_eager_imports()` doesn't say to me either deep or shallow imports, and it implies shallow imports because you have to explicitly name some module paths.  The other question is whether it's possible to *unset* eager imports you've specified in this call (or whether you even need to).  With the context manager at least, you understand the limited effective scope of the eagerness.

It's totally fine to be opinionated in the PEP and I support either interpretation. (It will likely get bikeshedded anyway :smile: .) The PEP should include a discussion of the options though, perhaps in the rejected ideas section.